### PR TITLE
[Feature] Temporal Pipeline Parallelism & Stream Batch for Real-Time Video

### PR DESCRIPTION
**Pull request by mnasser02**
## Purpose

Related issue: #2280. 

Tested on CausVid Wan2.1-T2V-1.3B #4441 (not merged yet).

**Implements temporal PP driven by micro-step execution, along with StreamDiffusionV2 features.**

- `SupportsMicroStepExecution` on `CausVidPipeline`
- `StreamAutoencoderKLWan`
- micro-step execution path with async comms, double-buffering, compute/comm overlap
- `StreamBatchScheduler` driving micro-step execution pipeline with stream-batch
- adaptive sink tokens and rolling kv cache per denoising slot
- motion-aware noise controller, RoPE offset reset
- dynamic scheduling of DiT blocks across PP stages, after taking into account vae encode/decode and DiT latencies.

**Not in scope:**
- streaming mode (currently, video is processed and generated offline)

## Test Plan

```
pytest -v \
tests/diffusion/test_diffusion_micro_step_pipeline.py \
tests/diffusion/test_diffusion_scheduler.py::TestStreamBatchScheduler
```

## Test Result

- **Unit tests:** 32 passed, 16 warnings

- **E2E on H100s**:

```
# KV cache config
local_attn_size = 3
sink_size = 3
sink_threshold = 0.2
```
### Throughput

image

### Per-rank occupancy  

image

### Bird example

[input prompt and video](https://github.com/chenfengxu714/StreamDiffusionV2/tree/master/examples)

<table>
  <tr>
    <td align="center"><b>1 step</b><br><video src="https://github.com/user-attachments/assets/089f9df2-1271-45af-ac11-0f8356a0480e
" controls width="100%"></video></td>
    <td align="center"><b>2 steps</b><br><video src="https://github.com/user-attachments/assets/df038625-f9a1-4736-869c-dced1e49461f" controls width="100%"></video></td>
  </tr>
  <tr>
    <td align="center"><b>3 steps</b><br><video src="https://github.com/user-attachments/assets/bfdc0e89-598b-44d8-ab48-1da3b5a4dff2" controls width="100%"></video></td>
    <td align="center"><b>4 steps</b><br><video src="https://github.com/user-attachments/assets/8c1ec6b9-e3c4-451b-8930-435da0bdfa64" controls width="100%"></video></td>
  </tr>
</table>

---
<details>
<summary> Essential Elements of an Effective PR Description Checklist </summary>

- [x] The purpose of the PR, such as "Fix some issue (link existing issues this PR will resolve)".
- [x] The test plan. Please provide the test scripts & test commands. Please state the reasons if your codes don't require additional test scripts. For test file guidelines, please check the [test style doc](https://docs.vllm.ai/projects/vllm-omni/en/latest/contributing/ci/tests_style/)
- [x] The test results. Please paste the results comparison before and after, or the e2e results.
- [ ] (Optional) The necessary documentation update, such as updating `supported_models.md` and `examples` for a new model. **Please run `mkdocs serve` to sync the documentation editions to `./docs`.**
- [ ] (Optional) Release notes update. If your change is user-facing, please update the release notes draft.
</details>

**BEFORE SUBMITTING, PLEASE READ <https://github.com/vllm-project/vllm-omni/blob/main/CONTRIBUTING.md>** (anything written below this line will be removed by GitHub Actions)

### CODE REVIEW

**Comment by hsliuustc0106** [comment]
Ready for full review when WIP status is removed. Preliminary scan available on request.

Note: Test plan and test results sections are currently empty. Please provide:
1. Test scripts & commands for the new pipeline parallelism and stream batch scheduler
2. E2e test results demonstrating the 40+ FPS target on 4×H100 with sub-500ms TTFF
3. Benchmark comparisons against SP baseline and PP×SP hybrid

**Comment by asukaqaq-s** [comment]
Hi, may I ask if there is an estimated plan for merging this PR? I have some features that need to be built on top of it.

**Comment by mnasser02** [comment]
@asukaqaq-s @fhfuih 

Currently, the core functionality of StreamDiffusionV2 (temporal PP through micro-step execution path and StreamBatchScheduler) should be almost ready. However, we can't test it properly since causal Wan isn't available in vllm-omni (in the paper, SDV2 builds on CausVid applied to Wan); not sure if this is being worked on, or this can be left for a future PR.

We have done some tests on a not-so-correct V2V adaptation of Wan2.1-T2V-1.3B to ensure functionality. I think that before the end of this week, we can open the PR (after a cleaning the code a bit) and share our preliminary results. 

Update: we decided to test on Lingbot #3701

**Comment by hsliuustc0106** [comment]
resolve conflicts please

File: vllm_omni/diffusion/models/lingbot_world_fast/fm_solvers_unipc.py [hunk 1]
```diff
@@ -0,0 +1,576 @@
+# Adapted from Lingbot-World/wan/utils/fm_solvers_unipc.py
+# Originally derived from https://github.com/huggingface/diffusers/blob/v0.31.0/src/diffusers/schedulers/scheduling_unipc_multistep.py
+# Converted to flow matching.
+
+
+import math
+
+import numpy as np
+import torch
+from diffusers.configuration_utils import ConfigMixin, register_to_config
```
Comment (wtomin) [inline vllm_omni/diffusion/models/lingbot_world_fast/fm_solvers_unipc.py:10]: ModelMixin / ConfigMixin should not be imported.

File: vllm_omni/diffusion/models/lingbot_world_fast/wan_fast.py [hunk 2]
```diff
@@ -0,0 +1,870 @@
+"""Some of the functions are borrowed from SelfForcing (https://github.com/guandeh17/Self-Forcing)."""
+
+import math
+
+import torch
+import torch.nn as nn
+import torch.nn.functional as torch_F
+from diffusers.configuration_utils import ConfigMixin, register_to_config
```
Comment (wtomin) [inline vllm_omni/diffusion/models/lingbot_world_fast/wan_fast.py:8]: Same here.

**Review by wtomin — CHANGES_REQUESTED** [review]
Merge conflict and readthedocs build need fixing before review proceeds. This PR is also substantial (10k+ LOC / 63 files); please run L3 tests locally and paste the results.

I also see new production files still importing diffusers mixins (ConfigMixin / ModelMixin). Please replace them with vLLM-Omni native abstractions rather than adding new diffusers dependencies in model code.
